docs: add Fully type-aware linting for Astro, Svelte, and Vue blog post (#2456)
flint-fyi/flint: packages/site/src/content/docs/blog/fully-type-aware-linting-for-astro-svelte-and-vue.mdx

Thanks to this, the type of every identifier used in the `@click` handler can be resolved to its actual value in tooling such as type-aware linting.

This is how [`vue-tsc`](https://github.com/vuejs/language-tools/tree/94907be4f056f25867e46a117ab18d2782b425d7/packages/tsc) type-checks `.vue` files internally!

## Using Volar.js for Type-Aware Linting

Flint's `@flint.fyi/volar-language` package allows creating TypeScript-based languages with full typed linting support.
In order to add support for a new language, a plugin author must provide:

- A list of custom file extensions
- Volar.js language plugin
- Optional extra properties that are passed to the rule context

Now, let's explore how this works from a user's perspective.

### Initialization

Initialization of a language happens implicitly.

Take a look at the example Flint config:

```typescript 'import "@flint.fyi/vue"'
// flint.config.ts
import { defineConfig } from "flint";
import "@flint.fyi/vue";

export default defineConfig({
	// ...
});
```

By importing `@flint.fyi/vue`, you enable _all TypeScript-only rules_ to support `.vue` linting.
No additional configuration needed!

> ### Lint rules

Reviewer: ```suggestion
### Lint Rules
```

1. `function vueComponent() { return MyComponent }` is not reported!
   The type of the `MyComponent` identifier is resolved to a valid Vue component type.
It's no longer an `any`.
2. `function any() { return anyValue }` is reported by [`ts/anyReturns`](/rules/ts/anyreturns).
This shows us that `return MyComponent` wasn't reported not because the rule is broken.
In fact, it's functioning as expected, reporting all `any`-typed values returned from functions.
3. `@click="() => { anyFunction() }"` in Vue `<template>` is reported by the [`ts/anyCalls`](/rules/ts/anycalls) rule.
We can see that the TypeScript-only rule works in JavaScript expressions embedded in Vue templates just fine!
4. `<div v-for="i in 3"></div>` is reported by `vue/vForKeys`.
Vue rules can access the Vue AST and analyze it.

## What's next

Flint is still in its early pre-alpha stages.
We're excited to show off this early technical preview of the novel architecture we're building into the linter.

The next steps for Flint's languages support are:

- Implement all Astro, Svelte, and Vue rules in their Flint plugins
- Investigate support for other TypeScript-based languages such as Ember and MDX

> 

Reviewer: 💡 🧠 you'd asked about what else to put here. I just realized some readers will probably wonder about general roadmaps. Like, if this project is so exciting and powerful, when can they try it out / why can't they yet? 

Maybe mentioning our next post in the blog will talk about timelines for this year and the next?

---
authors:
  name: auvred
  title: Maintainer
  picture: /team/auvred.webp
  url: https://github.com/auvred
date: 2026-03-18

> excerpt:

Reviewer: It's used for `<meta name="description" />`. I can verify locally that updates to it show up in the rendered page. Maybe this is docs gap in the Starlight Blog docs? Or just an ambiguity around pass-through properties?

Also, it's used in the og image generation:

https://github.com/flint-fyi/flint/blob/4b32a64faef08abf371fc0d5a40039f8660fa1bb/packages/site/src/pages/og/%5B...path%5D.ts#L18

Volar.js ships a nice VS Code extension called [Volar Labs](https://volarjs.dev/core-concepts/volar-labs/).
It will help us understand how custom syntax is translated into TS code.

![Volar.js virtual generated code](../../../assets/blog/flawless-type-aware-linting-for-astro-svelte-and-vue/volar-virtual-code.webp)

As you may notice, all embedded JS expressions, including code from the `<script>` block, are merged into a single TS file.
They're also transformed in order to be type-checkable.
For example, variables declared in `<script setup>` are prefixed with `__VLS_ctx.` when used in `<template>`.
Thanks to this, the type of every identifier used in the `@click` handler can be resolved to its actual value.

This is how `vue-tsc` type-checks `.vue` files internally!

## How Flint utilizes Volar.js for type-aware linting

The new `@flint.fyi/volar-language` package allows creating TS-based languages with ease.
In order to add support for a new language, a plugin author must provide:

- A list of custom file extensions
- Volar.js language plugin
- Optional extra properties that are passed to the rule context

We will explore three stages of linting.

> ### Initialization

Reviewer: Sure, that makes sense to me! I like the compromise 🙂 and think you're right, that's definitely a hugely important part of the story. And a great showcase of the APIs you set up!

> ### Initialization

Reviewer: Oh just clarifying my vague response: yes, I think moving the first two sections (Initialization, TS program creation) to architecture docs and leaving in the Linting section is the right way to go.

JSX-based frameworks like React and Solid.js natively work with the TypeScript compiler's `.jsx` and `.tsx` file support.
But more extensive frameworks like Astro, Solid, and Vue that add language-specific additions and non-standard file extensions need more tooling to work well with TypeScript.

Traditional linters such as ESLint have been able to work with those languages on a _syntax-only_ level.
They can use the languages' provided parsers to get an AST for the languages' contents and report errors purely based on that AST.

However, syntax-only linting is often not enough to catch all bugs.
[Type-aware linting](https://typescript-eslint.io/blog/typed-linting) allows for significantly more powerful lint rules -- but requires more deep integrations with TypeScript tooling.

We're excited to say that Flint is the first modern linter with full support for typed linting on extension languages.
Flint uses Volar.js to integrate TypeScript with extension languages, giving lint rules access to full type information even in embedded JavaScript expressions within language-specific files.

> This represents a big leap forward in linting for these languages.

Reviewer: (to go along with the earlier suggestion of moving this paragraph earlier)

```suggestion
Flint's goals include providing a complete linting experience out-of-the-box.
That includes full typed linting without any additional configuration.
To provide this for users, we needed to finally solve long-standing questions around how to fully type extension languages and their embedded code with TypeScript.
```

> ## `no-unsafe-*` rules and embedded JavaScript expressions

Reviewer: I think this goes along with what we'd agreed, but please lmk if not. It's that the framing is around typed rules in general, but we want to call out `no-unsafe-*` specifically?

```suggestion
## Typed Rules and Embedded JavaScript Expressions
```

This applies to all type-aware rules, but `@typescript-eslint/no-unsafe-*` rules break more often than others.
The problem is that imports from files with custom extensions aren't supported by TypeScript natively.
So all imports from [`.astro`](https://github.com/ota-meshi/eslint-plugin-astro/issues/348), [`.svelte`](https://github.com/sveltejs/eslint-plugin-svelte/issues/1150), or [`.vue`](https://github.com/vuejs/vue-eslint-parser/issues/104) files are resolved to `any` when files are parsed using `@typescript-eslint/parser`.

:::note
[Yosuke Ota](https://github.com/ota-meshi)'s [`typescript-eslint-parser-for-extra-files`](https://github.com/ota-meshi/typescript-eslint-parser-for-extra-files) made progress on providing type information on TypeScript-based language files in ESLint.
See the library's documentation for more issues and excellent prior art.
:::

Another major user-facing problem is that embedded JavaScript expressions in extension languages can't be linted with type information.

> For example, traditional linters would not be able to determine type information in the following Vue snippet to know whether the `log(hello);` call is safe.

Reviewer: [Phrasing] I commonly prefer a >=3 sentence flow: a setup, an explainer, and a lead to the next thing. This has a setup and a lead but I found lacking an explainer.

```suggestion
Another major user-facing problem is that embedded JavaScript expressions in extension languages can't be linted with type information.
TypeScript cannot natively type-check syntax outside of JavaScript, JSX, or TypeScript.
For example, traditional linters would not be able to use TypeScript to determine type information in the following Vue snippet to know whether the `log(hello);` call is safe.
```

:::note
Vue is used in illustrations and examples throughout this post, but similar logic and rules apply for all other TypeScript-based languages.
:::

```vue {11-13} /(?<!console.)log/ "hello"
<script lang="ts" setup>
	function log(msg: string) {
		console.log(msg);
	}
	const hello = "Hello world!";
</script>

<template>
	<button
		@click="
			() => {
				log(hello);
			}
		"
	>
		Click me!
	</button>
</template>
```

Notice how the `@click` event listener is a JavaScript expression that uses variables declared in `<script>`.
In order to lint this JavaScript expression _properly_, we need the ability to resolve the types used inside it.

Resolving types in extension languages is a difficult technical challenge.
Unfortunately, it's not solveable by straightforward techniques like concatenating all JavaScript expressions into one big file.
Embedded JavaScript expressions rely on many language-specific quirks that must be taken into account.

Take, for example, the following code that uses a custom Vue component:

```vue "data"
<template>
	<MyComponent v-slot="{ data }">
		{{ data.text }}
	</MyComponent>
</template>
```

`data` comes from `MyComponent`'s `v-slot`.

> To know its type, we must resolve the type of `MyComponent`, extract its `v-slot` type, and so on.

Reviewer: ```suggestion
To know its type, we must resolve the type of `MyComponent`, extract its `v-slot` type, and resolve its type based on Vue-specific logic.
```

This is why naive concatenation wouldn't work.

[Johnson Chu](https://github.com/johnsoncodehk)'s [`tsslint`](https://github.com/johnsoncodehk/tsslint) ("The lightest TypeScript semantic linting solution in JS") used Volar.js to solve this problem, though the first problem (`any`-typed imports) is still present.
It supports loading Volar.js language plugins for Astro, MDX, Vue, and Vue Vine.
Thanks to this, it supports type-aware linting for these languages.
However, it lacks the TypeScript patching mechanism required to resolve custom file extensions to concrete types instead of `any`.

> It also does not expose custom language ASTs to lint rules.

Reviewer: [Phrasing]  +1 to the shouting out tsslint and what+how it does, I think this is a good positive note. I want to be very careful that we about accidentally coming across as implying anything negative / _"they don't do this good"_ about them. Just to be extra causious I'd phrase this as _"tsslint was a start with X, here's how we continue things with Y"_ and not _"tsslint does X things well, but lacks in Y"_.

```suggestion
[Johnson Chu](https://github.com/johnsoncodehk)'s [`tsslint`](https://github.com/johnsoncodehk/tsslint) ("The lightest TypeScript semantic linting solution in JS") used Volar.js to solve the problem of type-aware linting for language plugins.
It achieved support for loading Volar.js language plugins for Astro, MDX, Vue, and Vue Vine.
In doing so, `tsslint` solved one of the gaps in typed linting: linting extension language files on their own.

Flint builds on `tsslint`'s feature set by additionally solving the first problem  of `any`-typed imports from extension language files.
Flint  includes a dynamic TypeScript patching mechanism required to resolve custom file extensions to concrete types instead of `any`.
It also exposes custom language ASTs to lint rules, which allows lint rules to fully operate on both native TypeScript files as well as extension language files.
```

...roughly?

Reviewer: I also think it'd be good to ping Johnson Chu about this just to make sure we're not conflicting with how they want tsslint to be phrased. (I try to do this for all explicit references of neighboring projects)

## Type Checking TypeScript-Based Extension Languages

> At some point, every TypeScript-based-language author wants to type-check their language with TypeScript.

Reviewer: [Phrasing] Nit: it's not just the author type-checking their own code! This phrasing expands the meaning to be everyone:

```suggestion
At some point, every TypeScript-based-language author wants to enable type-checking their language with TypeScript.
```

2. Resolving custom file extensions in TypeScript's existing module resolution system (so they are not resolved to `any`)

Even though TypeScript can be extended with [Language Service Plugins](https://github.com/microsoft/TypeScript/wiki/Writing-a-Language-Service-Plugin), these plugins can't add new custom syntax to TypeScript.
Moreover, plugins extend the _editing experience only_, i.e., they can't be used in CLI workflows, such as `tsc --noEmit`.

That's why TypeScript-based-language authors were required to invent hacks that could force TypeScript to understand their custom languages.

Several different approaches to this problem exist.
For example, [`svelte-check`](https://github.com/sveltejs/language-tools/tree/master/packages/svelte-check) and [`ngtsc`](https://github.com/angular/angular/tree/main/packages/compiler-cli/src/ngtsc) both use TypeScript compiler to type-check custom syntax.
However, they use different approaches, which makes it difficult to unify them within a single linting architecture.

Thankfully, there is a tool that can provide unified type-checking for custom languages.

## Volar.js

> 

Reviewer: [Phrasing] Whenever a new section with an important new subject/area is started, it's good to define the subject. Newcomers reading this for the first time don't yet know what Volar.js is beyond "it's something that helps with the problem statement". They don't yet have a reason to care about its history.

I quite like the later paragraph that does actually define Volar. How about bringing it up here?

```suggestion

[Volar.js](https://volarjs.dev), the "Embedded Language Tooling Framework", is a framework that provides language authors with a high-level scaffolding that handles all the complexities of language tooling.
It handles such as LSP integration, routing of LSP requests to the appropriate embedded language service, type-checking powered by TypeScript, and much more -- so that authors can avoid each reimplementing those areas of tooling.

```

It all started as part of the Vue Language Tools initiative!

> [Johnson Chu](https://github.com/johnsoncodehk/) developed the outstanding language support for Vue.js, and then they noticed the pattern that could be decoupled from Vue Language Tools to support almost any possible TypeScript-based language.

Reviewer: wowow. heroic indeed! (in that case, 🔪 the "and team" to give Johnson the heroic level of credit)

This is how Volar.js, the "Embedded Language Tooling Framework", was created.

What does "embedded" mean?

Well, take for example a Vue Single File Component (SFC).
This file contains blocks with three different languages: TypeScript (`<script lang="ts">`), HTML (`<template>`), and CSS (`<style lang="css">`).

> The HTML block also has an embedded TypeScript/JavaScript expression inside it.

Reviewer: [Phrasing] At this point, I think most readers will think they already know what "embedded" means. The post has explained it as the concept of an embedded snippet. What I think will really drive home the point & integrate it with Volar is explaining how it integrates in the high-level Volar architecture/usage:

```suggestion
### Embedded Languages Support

A key benefit of using Volar.js is its full support for various forms of embedded languages.
It allows language tooling authors to define various forms of embedded syntax, as well as how types flow through that syntax.

The following code block shows an archetypal Vue Single File Component (SFC).
It contains three different embedded languages: TypeScript (`<script lang="ts">`), HTML (`<template>`), and CSS (`<style lang="css">`).
The HTML block also has an embedded TypeScript/JavaScript expression inside it.
```

[90m╭[39m[38;2;255;119;119m./[39m[1m[38;2;255;73;73msrc/index.ts[39m[22m
[90m│ [39m
[90m│ [39m[38;2;238;170;119m[38;2;204;170;170m[[39m[38;2;238;170;119m[38;2;225;103;103m[1mts/anyReturns[22m[39m[38;2;238;170;119m[38;2;204;170;170m][39m[38;2;208;140;89m Unsafe return of a value of type `any`.[39m
[90m│ [39m
[90m│ [39m[38;2;147;147;147m6:18 [39m[90m│[39m function any() { return anyValue }
[90m│ [39m[90m     │ [39m[38;2;255;73;73m                 ~~~~~~~~~~~~~~~[39m
[90m│ [39m
[90m│ [39m [38;2;154;137;120m[3mReturning a value of type `any` or a similar unsafe type defeats TypeScript's type safety guarantees.[23m[39m
[90m│ [39m [38;2;154;137;120m[3mThis can allow unexpected types to propagate through your codebase, potentially causing runtime errors.[23m[39m
[90m│ [39m
[90m│ [39m [38;2;117;168;185mSuggestion: [38;2;187;204;221m[38;2;137;154;171mEnsure the returned value has a well-defined, specific type.[39m[38;2;187;238;255m[38;2;187;204;221m[39m[38;2;187;238;255m[39m
[90m│ [39m
[90m│ [39m [38;2;130;164;130m[3m→ flint.fyi/rules/ts/anyreturns[23m[39m
[90m│ [39m

> [90m╰[39m

Reviewer: Here and the other `╰`. Unless, did you omit this intentionally?

```suggestion
[90m╰──────────────────────────────────────────────────────────────────────────[39m
```

[90m│ [39m
[90m│ [39m[38;2;238;170;119m[38;2;204;170;170m[[39m[38;2;238;170;119m[38;2;225;103;103m[1mts/vForKeys[22m[39m[38;2;238;170;119m[38;2;204;170;170m][39m[38;2;208;140;89m Elements using v-for must include a unique :key to ensure correct reactivity and DOM stability.[39m
[90m│ [39m
[90m│ [39m[38;2;147;147;147m 8:8 [39m[90m│[39m <div v-for="i in 3"></div>
[90m│ [39m[90m     │ [39m[38;2;255;73;73m     ~~~~~[39m
[90m│ [39m
[90m│ [39m [38;2;154;137;120m[3mA missing :key can cause unpredictable updates during rendering optimizations.[23m[39m
[90m│ [39m [38;2;154;137;120m[3mWithout a key, Vue may reuse or reorder elements incorrectly, which breaks expected behavior in transitions and stateful components.[23m[39m
[90m│ [39m
[90m│ [39m [38;2;117;168;185mSuggestion: [38;2;187;204;221m[38;2;137;154;171mAlways provide a unique :key based on the v-for item, such as an id.[39m[38;2;187;238;255m[38;2;187;204;221m[39m[38;2;187;238;255m[39m
[90m│ [39m
[90m│ [39m [38;2;130;164;130m[3m→ flint.fyi/rules/vue/vforkeys[23m[39m
[90m│ [39m
[90m╰[39m

[31m✖ Found [1m3 reports[22m across [1m2 files[22m.[39m

> ```

Reviewer: [Non-Actionable] Just cross-linking #1287 for reference. That'd be nice. I'll ping @yonava; I know they've got a lot going on these days.
